### PR TITLE
MEPTS-325: change request in monthly report resumo mensal

### PR DESCRIPTION
**Pull request by ningosi**
Changes requested on MEPTS-325

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java [hunk 1]
```diff
@@ -359,4 +359,76 @@ public static String getF3Exclusion(int encounterType) {
...
+      int mastercardEncounterType, int consultationEncounterType, int preArtStarConceptId) {
+    String query =
+        "SELECT l.patient_id FROM"
+            + " (SELECT p.patient_id AS patient_id, MIN(e.encounter_datetime) AS encounter_datetime FROM patient p INNER JOIN encounter e ON p.patient_id=e.patient_id"
+            + " INNER JOIN(SELECT p.patient_id AS patient_id,o.value_datetime AS art_start_date FROM patient p INNER JOIN encounter e ON p.patient_id=e.patient_id"
+            + " INNER JOIN obs o ON o.encounter_id=e.encounter_id WHERE p.voided=0 AND e.voided=0 AND o.voided=0 AND e.encounter_type=%d AND e.location_id=:location "
+            + " AND o.value_datetime IS NOT NULL AND o.concept_id=%d GROUP BY p.patient_id) pre_art ON p.patient_id=pre_art.patient_id WHERE p.voided=0 "
+            + " AND e.voided=0 AND e.encounter_type=%d AND e.location_id=:location AND encounter_datetime <= :endDate  AND encounter_datetime=pre_art.art_start_date GROUP BY patient_id) l";
+    return String.format(
+        query, mastercardEncounterType, consultationEncounterType, preArtStarConceptId);
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: Mappings for the last two format strings are switched, query expects a `concept_id` then an `encounter_type_id`.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java [hunk 2]
```diff
@@ -467,4 +467,36 @@ public CohortDefinition getPatientsLostToFollowUpAndNotDeadTransferredOrStoppedT
...
     return cd;
   }
+
+  /**
+   * Get patients who have encounter of a specific type within date boundaries
+   *
+   * @param encounterType
+   * @retrun CohortDefinition
+   */
+  public CohortDefinition getPatientsHavingEncounterWithinDateBoundaries(int encounterType) {
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java:460]: Why not use `EncounterCohortDefinition` instead?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java [hunk 3]
```diff
@@ -836,4 +863,62 @@ public CohortDefinition getNumberOfPatientsWithAtLeastOneClinicalAppointmentDuri
...
+  public CohortDefinition getNumberOfPatientsWhoAbandonedArtDuringCurrentMonthForB7() {
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Number of patients who Abandoned the ART during the current month");
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.setQuery(
+        ResumoMensalQueries.getNumberOfPatientsWhoAbandonedArtDuringCurrentMonthB7(
+            hivMetadata.getARVPharmaciaEncounterType().getEncounterTypeId(),
+            hivMetadata.getMasterCardEncounterType().getEncounterTypeId(),
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java]: It should be `hivMetadata.getMasterCardDrugPickupEncounterType()` (encounter_type 52).

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java [hunk 4]
```diff
@@ -566,8 +593,8 @@ public CohortDefinition getStandardDefinitionForEcolumns() {
     cd.addSearch(
         "B7",
         map(
-            getNumberOfPatientsWhoAbandonedArtDuringCurrentMonthB7(),
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java:569]: Old B7 CohortDefinition is only used in a unit test, maybe it should be deleted.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java [hunk 5]
```diff
@@ -836,4 +863,62 @@ public CohortDefinition getNumberOfPatientsWithAtLeastOneClinicalAppointmentDuri
...
+            hivMetadata.getArtDatePickup().getConceptId()));
+    return cd;
+  }
+
+  /**
+   * B12-B7A: Number of active patients in ART by end of previous month
+   *
+   * @retrun CohortDefinition
+   */
+  public CohortDefinition getNumberOfPatientsWhoAbandonedArtDuringPreviousMonthForB127A() {
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java:884]: This query does not seem necessary. We can always use `getNumberOfPatientsWhoAbandonedArtDuringCurrentMonthForB7` changing the parameter mappings.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java [hunk 6]
```diff
@@ -359,4 +359,76 @@ public static String getF3Exclusion(int encounterType) {
...
+   * @return String
+   */
+  public static String getNumberOfPatientsWhoAbandonedArtDuringPreviousMonthB127A(
+      int pharmacyEncounterType, int mastercardEncounterType, int drugPickupConceptId) {
+    String query =
+        "SELECT final.patient_id FROM( "
+            + " SELECT c.patient_id, MAX(c.encounter_date) as encounter_date FROM( "
+            + " SELECT p.patient_id, MAX(e.encounter_datetime) AS encounter_date FROM patient p INNER JOIN encounter e "
+            + " ON p.patient_id=e.patient_id WHERE p.voided=0 AND e.voided=0 AND e.location_id=:location AND "
+            + " e.encounter_datetime <=:endDate AND e.encounter_type=%d GROUP BY p.patient_id "
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: Spec says by reporting `startDate`.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java [hunk 7]
```diff
@@ -359,4 +359,76 @@ public static String getF3Exclusion(int encounterType) {
...
+            + " UNION  "
+            + " SELECT p.patient_id, MAX(o.value_datetime) AS encounter_date FROM patient p INNER JOIN encounter e "
+            + " ON p.patient_id=e.patient_id INNER JOIN obs o ON o.encounter_id=e.encounter_id WHERE p.voided=0 AND e.voided=0 "
+            + " AND o.voided=0 AND e.encounter_type=%d AND o.concept_id=%d AND o.value_datetime IS NOT NULL AND "
+            + "e.location_id=:location AND e.encounter_datetime<=:endDate GROUP BY p.patient_id "
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: Spec says by reporting `startDate`.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java [hunk 8]
```diff
@@ -836,4 +863,62 @@ public CohortDefinition getNumberOfPatientsWithAtLeastOneClinicalAppointmentDuri
...
+   * B7: Number of patientes who Abandoned the ART during the current month
+   *
+   * @retrun CohortDefinition
+   */
+  public CohortDefinition getNumberOfPatientsWhoAbandonedArtDuringCurrentMonthForB7() {
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Number of patients who Abandoned the ART during the current month");
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java]: The query does not expect a `startDate` parameter.

**Review by ynurmahomed — CHANGES_REQUESTED** [review]
Compositions all look to be fine. There are some issues with the queries, please have a look.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java (hunk 1 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: Thanks, this is corrected

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java (hunk 2 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java:460]: We have had discrepancies in the numbers we get by using the reporting ORM against the native queries, but I agree we can utilize what already exists other that duplication. I will change to use that inbuilt classes from reporting framework.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java (hunk 3 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java]: Fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java (hunk 4 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java:569]: This was removed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java (hunk 6 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: StartDate is applied to the overal query after adding 90 days

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java (hunk 7 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: same as above

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java (hunk 6 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: @ningosi I'm refering to where the spec says:

> 1. Select the most recent date:
>     1. Last encounter_datetime from encounter type 18 by reporting **start date**

Please have another look.

Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: You are right. I will change this

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java (hunk 2 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java]: @ningosi this CohortDefinition has no usages now

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java (hunk 4 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java:569]: @ningosi the method is still present in the `ResumoMensalCohortQueries` class.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java (hunk 7 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/ResumoMensalQueries.java]: @ningosi from the spec:

> The most recent “Data de Levantamento” (Concept Id 23866 value_datetime-not empty/null) by reporting **start date** from encounters of type 52

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java (hunk 2 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java:460]: @ningosi sql query is still present.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java (hunk 4 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java:569]: fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java [hunk 9]
```diff
@@ -836,4 +855,43 @@ public CohortDefinition getNumberOfPatientsWithAtLeastOneClinicalAppointmentDuri
...
+  public CohortDefinition getNumberOfPatientsWhoAbandonedArtDuringPreviousMonthForB127A() {
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Number of patients who Abandoned the ART during the current month");
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.setQuery(
+        ResumoMensalQueries.getNumberOfPatientsWhoAbandonedArtDuringPreviousMonthB127A(
+            hivMetadata.getARVPharmaciaEncounterType().getEncounterTypeId(),
+            hivMetadata.getMasterCardEncounterType().getEncounterTypeId(),
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java]: @ningosi it should be `hivMetadata.getMasterCardDrugPickupEncounterType().getEncounterTypeId()`

**Review by ynurmahomed — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java (hunk 5 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java:884]: @ningosi it now only needs parameters `location` and `startDate`. Maybe `startDate` name could be changed to something like `onDate` for better understanding.

**Review by ynurmahomed — APPROVED** [review]

**Pull request merged**